Source: Written by Michael Forney
Reason: Use supplied terminal interface (if given), rather than hardcoding some
        defaults.

--- a/crypto/ui/ui_openssl.c	2013-02-11 07:26:04.000000000 -0800
+++ b/crypto/ui/ui_openssl.c	2013-09-29 01:01:57.205644041 -0700
@@ -184,22 +184,12 @@
  * TERMIO, TERMIOS, VMS, MSDOS and SGTTY
  */
 
-#if defined(__sgi) && !defined(TERMIOS)
+#if defined(__sgi)
 # define TERMIOS
-# undef  TERMIO
-# undef  SGTTY
-#endif
-
-#if defined(linux) && !defined(TERMIO)
-# undef  TERMIOS
-# define TERMIO
-# undef  SGTTY
 #endif
 
-#ifdef _LIBC
-# undef  TERMIOS
+#if (defined(linux) || defined(_LIBC))
 # define TERMIO
-# undef  SGTTY
 #endif
 
 #if !defined(TERMIO) && !defined(TERMIOS) && !defined(OPENSSL_SYS_VMS) && !defined(OPENSSL_SYS_MSDOS) && !defined(OPENSSL_SYS_MACINTOSH_CLASSIC) && !defined(MAC_OS_GUSI_SOURCE)
@@ -220,6 +210,14 @@
 #undef SGTTY
 #endif
 
+/* Prefer TERMIOS, then TERMIO, then SGTTY */
+#if defined(TERMIOS)
+# undef TERMIO
+# undef SGTTY
+#elif defined(TERMIO)
+# undef SGTTY
+#endif
+
 #ifdef TERMIOS
 # include <termios.h>
 # define TTY_STRUCT		struct termios
